crates/worker/src/purgatory.rs

```diff
@@ -171,3 +195,80 @@ impl Inner {
         });
     }
 }
+
+struct GlobalState {
+    db_client: DynamoDBClient,
+    s3_client: S3Client,
+    pub items: Vec<Item>,
+}
+
+impl GlobalState {
+    pub fn new(db_client: DynamoDBClient, s3_client: S3Client) -> Self {
+        Self {
+            db_client,
+            s3_client,
+            items: vec![],
+        }
+    }
+
+    pub async fn sync(&mut self, sync_from: &DateTime<Utc>) -> Result<(), GlobalPurgeError> {
+        const MAX_CAPACITY: usize = 1000;
+
+        let mut last_evaluated_key = None;
+        loop {
+            let output = self
+                .db_client
+                .client
+                .scan()
+                .table_name(self.db_client.table_name.clone())
+                .filter_expression("CreatedAt <= :created_at")
+                .expression_attribute_values(
+                    ":created_at",
+                    AttributeValue::S(sync_from.to_rfc3339()),
+                )
+                .limit(MAX_CAPACITY as i32)
+                .set_exclusive_start_key(last_evaluated_key)
+                .send()
+                .await
+                .unwrap();
+
+            let raw_items = if let Some(items) = output.items {
+                items
+            } else {
+                break Ok(());
+            };
+
+            let remaining_capacity = MAX_CAPACITY - self.items.len();
+            let mut to_append: Vec<Item> = raw_items
+                .into_iter()
+                .take(remaining_capacity)
+                .map(|raw_item| raw_item.try_into())
+                .collect::<Result<_, ItemError>>()?;
+
+            self.items.append(&mut to_append);
+            if self.items.len() == MAX_CAPACITY {
+                break Ok(());
+            }
+
+            last_evaluated_key = if let Some(last_evaluated_key) = output.last_evaluated_key {
```

Comment:
Shouldn't last_evaluated be retrieved from `to_append` vector?

Reply:
Not really, to_append contains el-ts that satisfy condition, while dynamodb could analyze more el-ts than that. Refer [here](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html)